### PR TITLE
refactor dependencies again, rebuild module dependencies

### PR DESCRIPTION
**Pull request by duncathan**

### CODE REVIEW

File: retro_data_structures/formats/mrea.py [hunk 1]
```diff
@@ -60,6 +61,81 @@ class MREAVersion(IntEnum):
...
+            "dependencies" / PrefixedArray(Int32ub, Struct(
+                asset_id=asset_id,
+                asset_type=FourCC,
+            )),
+            "offsets" / PrefixedArray(Int32ub, Int32ub),
+        ))
+
+    def _decode(self, obj, context, path) -> AreaDependencies:
+        layers = []
+        for i in range(len(obj.offsets) - 1):
```
Comment (henriquegemignani) [inline retro_data_structures/formats/mrea.py:77]:
```python
for start, finish in zip(obj.offsets, obj.offsets[1:]):`
```
Maybe that?

File: retro_data_structures/formats/mrea.py [hunk 2]
```diff
@@ -60,6 +61,81 @@ class MREAVersion(IntEnum):
...
+
+class AreaModuleDependencyAdapter(Adapter):
+    def __init__(self):
+        super().__init__(Struct(
+            rel_module=PrefixedArray(Int32ub, String),
+            rel_offset=PrefixedArray(Int32ub, Int32ub),
+        ))
+
+    def _decode(self, obj, context, path) -> list[list[str]]:
+        offsets = list(zip(*[iter(obj.rel_offset)]*2))
```
Comment (henriquegemignani) [inline retro_data_structures/formats/mrea.py:119]:
```python
    offset_iter = iter(obj.rel_offset)
    return [
        obj.rel_module[start:finish]
        for start, finish in zip(offset_iter, offset_iter)
    ]
```

and you complain about my poor ternary :P

File: retro_data_structures/formats/mrea.py [hunk 3]
```diff
@@ -525,6 +601,26 @@ def generated_objects_layer(self) -> ScriptLayer:
...
+@dataclasses.dataclass(frozen=True)
+class AreaDependencies:
+    layers: list[list[Dependency]]
+    non_layer: list[Dependency]
+
+    @property
+    def all_dependencies(self) -> Iterator[Dependency]:
+        for layer in self.layers:
+            yield from layer
+        yield from self.non_layer
```
Comment (henriquegemignani) [inline retro_data_structures/formats/mrea.py:613]: Add some tests for this all plz

File: retro_data_structures/formats/mrea.py [hunk 4]
```diff
@@ -711,45 +797,91 @@ def build_mlvl_dependencies(self, only_modified: bool = False):
...
-        for layer_dep in layer_deps:
-            offsets.append(offset)
-            offset += len(layer_dep)
+        for instance in self.generated_objects_layer.instances:
+            layer = instance.id.layer
+            if (
+                (not only_modified)
+                or layers[layer].is_modified()
+            ):
+                layer_rels[layer].extend(instance.get_properties().modules())
```
Comment (henriquegemignani) [inline retro_data_structures/formats/mrea.py:840]: `modules` is a class method, this can be `instance.type.modules()` and be faster

**Review by henriquegemignani — CHANGES_REQUESTED** [review]

File: retro_data_structures/formats/mrea.py (hunk 2 above)
Comment (duncathan) [reply retro_data_structures/formats/mrea.py:119]: i *literally* took this from the python docs: https://docs.python.org/3/library/functions.html#zip

(i will change it)

**Pull request merged**